crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -32,6 +32,56 @@ def role_constraints
     end
   end
 
+  def shutdown_openstack_at_nodes
+    db_nodes = []
+    NodeObject.all.each do |node|
+      next unless node.state == "crowbar_upgrade"
+
+      # Find the nodes with Database role
+      # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade)
+      node.crowbar["run_list_map"].each do |role, v|
+        db_nodes.push node.name if role == "database-server"
+      end
+
+      # mark the position in the upgrade process
+      node["crowbar_wall"]["crowbar_openstack_shutdown"] = true
+      node.save
+    end
+
+    # Adapt current proposal, so only non-db nodes have crowbar-upgrade role
+    # (crowbar-upgrade role is currently assigned to all nodes)
+    proposal = ProposalObject.find_proposal('crowbar', 'default')
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] -= db_nodes
+    proposal.save
+
+    # Commit proposal so the shutdown actions from crowbar-upgrade get executed for non-db nodes
+    proposal_commit("default", false, false)
+
+    # After all non-DB services are shut down, adapt the proposal again and
+    # commit it with DB nodes only (and the special role for them)
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] = []
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-db-dump"] = db_nodes
+    proposal.save
+
+    # FIXME this proposal could return some error if there's not enough space for DB dump
```

Comment:
Style/CommentAnnotation: Annotation keywords like `FIXME` should be all upper case, followed by a colon, and a space, then a note describing the problem. (https://github.com/bbatsov/ruby-style-guide#annotate-keywords)